llvm/lib/Target/AIE/AIEWawRegRewriter.cpp

```diff
@@ -0,0 +1,455 @@
+//===----- AEIWaWRegRewriter.cpp - Rewrite Register to not include multiple Defs
+//---------===//
+//
+// This file is licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// (c) Copyright 2024 Advanced Micro Devices, Inc. or its affiliates
+//
+//===----------------------------------------------------------------------===//
+//
+// Implementation of a Loop Register rewriting pass, that reduces WAW physical
+// register conflicts, so that software pipelining compact loops better.
+//
+//===----------------------------------------------------------------------===//
+
+#include "AIEBaseInstrInfo.h"
+#include "AIEBaseRegisterInfo.h"
+#include "Utils/AIELoopUtils.h"
+
+#include "llvm/ADT/MapVector.h"
+#include "llvm/ADT/SmallSet.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/CodeGen/LiveDebugVariables.h"
+#include "llvm/CodeGen/LiveIntervals.h"
+#include "llvm/CodeGen/LiveRegMatrix.h"
+#include "llvm/CodeGen/MachineFunctionPass.h"
+#include "llvm/CodeGen/MachineInstr.h"
+#include "llvm/CodeGen/MachineInstrBuilder.h"
+#include "llvm/CodeGen/MachineLoopInfo.h"
+#include "llvm/CodeGen/MachineOperand.h"
+#include "llvm/CodeGen/Passes.h"
+#include "llvm/CodeGen/SlotIndexes.h"
+#include "llvm/CodeGen/TargetInstrInfo.h"
+#include "llvm/CodeGen/TargetSubtargetInfo.h"
+#include "llvm/MC/MCRegister.h"
+#include "llvm/Support/Debug.h"
+#include "llvm/Support/raw_ostream.h"
+#include <llvm/CodeGen/MachineBasicBlock.h>
+#include <map>
+
+using namespace llvm;
+
+#define DEBUG_TYPE "aie-wawreg-rewrite"
+
+namespace {
+
+cl::opt<bool> DisableCopySubRegHandling(
+    "aie-waw-disable-copy-sub-reg-handling",
+    cl::desc("Disable Copy Sub register handling, where WAW regs will not be "
+             "replaced, if reg parts are reused."),
+    cl::init(false), cl::Hidden);
```

Comment:
Could you add a specific test for that option? For example, we can add two different `RUN:` lines, with different `--check-prefix` FileCheck commands. I'm curious to see what case we are avoiding here. And it's generally it's great to have tests for each "special case" :)

Reply:
I added a run with the prefix disabled. 
Should I denote somewhere that redundant_sub_reg_moves is the function that tests this option?